service/service.go

```diff
@@ -0,0 +1,493 @@
+	ErrVerbNotSupported = errors.New("unsupported verb")
+)
+
+func (s Verb) String() string {
+	switch s {
+	case PingVerb:
+		return "PING"
+	case StatsVerb:
+		return "STATS"
+	case InfoVerb:
+		return "INFO"
+	case SchemaVerb:
+		return "SCHEMA"
+	default:
+		return ""
+	}
+}
+
+// Add adds a microservice.
+// It will enable internal common services (PING, STATS, INFO and SCHEMA) as well as
+// the actual service handler on the subject provided in config.Endpoint
+// A service name and Endpoint configuration are required to add a service.
+// Add returns a [Service] interface, allowing service menagement.
+// Each service is assigned a unique ID.
+func Add(nc *nats.Conn, config Config) (*Service, error) {
+	if err := config.valid(); err != nil {
+		return nil, err
+	}
+
+	id := nuid.Next()
+	svc := &Service{
+		Config: config,
+		conn:   nc,
+		id:     id,
+	}
+	svc.verbSubs = make(map[string]*nats.Subscription)
+	svc.endpointStats = make(map[string]*EndpointStats)
+	svc.endpointStats[""] = &EndpointStats{
+		Name: config.Name,
+	}
+
+	// Setup internal subscriptions.
+	var err error
+
+	svc.reqSub, err = nc.QueueSubscribe(config.Endpoint.Subject, QG, func(m *nats.Msg) {
+		svc.reqHandler(&Request{Msg: m})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	info := &Info{
+		Name:        config.Name,
+		ID:          id,
+		Description: config.Description,
+		Version:     config.Version,
+		Subject:     config.Endpoint.Subject,
+	}
+
+	ping := &Ping{
+		Name: config.Name,
+		ID:   id,
+	}
+
+	infoHandler := func(m *nats.Msg) {
+		response, _ := json.Marshal(info)
+		m.Respond(response)
+	}
+
+	pingHandler := func(m *nats.Msg) {
+		response, _ := json.Marshal(ping)
+		m.Respond(response)
+	}
+
+	statusHandler := func(m *nats.Msg) {
+		response, _ := json.Marshal(svc.Stats())
+		m.Respond(response)
+	}
+
+	schemaHandler := func(m *nats.Msg) {
+		response, _ := json.Marshal(svc.Schema)
+		m.Respond(response)
+	}
+
+	if err := svc.verbHandlers(nc, InfoVerb, infoHandler); err != nil {
+		return nil, err
+	}
+	if err := svc.verbHandlers(nc, PingVerb, pingHandler); err != nil {
+		return nil, err
+	}
+	if err := svc.verbHandlers(nc, StatsVerb, statusHandler); err != nil {
+		return nil, err
+	}
+
+	if svc.Schema.Request != "" || svc.Schema.Response != "" {
+		if err := svc.verbHandlers(nc, SchemaVerb, schemaHandler); err != nil {
+			return nil, err
+		}
+	}
+
+	svc.natsHandlers.closed = nc.Opts.ClosedCB
+	if nc.Opts.ClosedCB != nil {
+		nc.SetClosedHandler(func(c *nats.Conn) {
+			svc.Stop()
+			svc.natsHandlers.closed(c)
+		})
+	} else {
+		nc.SetClosedHandler(func(c *nats.Conn) {
+			svc.Stop()
+		})
+	}
+
+	svc.natsHandlers.asyncErr = nc.Opts.AsyncErrorCB
+	if nc.Opts.AsyncErrorCB != nil {
+		nc.SetErrorHandler(func(c *nats.Conn, s *nats.Subscription, err error) {
+			if config.ErrorHandler != nil {
+				config.ErrorHandler(svc, &NATSError{
+					Description: err.Error(),
+					Subject:     s.Subject,
+				})
+			}
+			svc.Stop()
```

Comment:
should this one be run in a goroutine? What could be interesting here is to check whether the error is a SlowConsumer error for example to have it drain, and maybe let the service resubscribe once its buffers have caught up.

Reply:
I'm not sure whether it should run in a goroutine... It is added in async handler anyway, I would probably keep it synchronous within one callback.

As to detecting e.g. slow consumer error (maybe others), I would do that in a separate PR if that's ok, just to make this one works properly and then add improvements.